homeassistant/components/viam/__init__.py

```diff
@@ -0,0 +1,330 @@
+            supports_response=SupportsResponse.ONLY,
+        )
+
+    def unload(self) -> None:
+        """Clean up any open clients."""
+        self.hass.services.remove(DOMAIN, "capture_data")
+        self.hass.services.remove(DOMAIN, "capture_image")
+        self.hass.services.remove(DOMAIN, "get_classifications")
+        self.hass.services.remove(DOMAIN, "get_detections")
+        self.client.close()
+
+    async def capture_data(self, call: ServiceCall) -> None:
+        """Accept input from service call to send to Viam."""
+        parts: list[RobotPart] = await self.client.app_client.get_robot_parts(
+            robot_id=self.data["robot_id"]
+        )
+        values = [call.data.get("values")]
+        component_type = call.data.get("component_type", "sensor")
+        component_name = call.data.get("component_name")
+
+        await self.client.data_client.tabular_data_capture_upload(
+            tabular_data=values,
+            part_id=parts.pop().id,
+            component_type=component_type,
+            component_name=component_name,
+            method_name="capture_data",
+            data_request_times=[(datetime.now(), datetime.now())],
+        )
+
+    async def capture_image(self, call: ServiceCall) -> None:
+        """Accept input from service call to send to Viam."""
+        parts: list[RobotPart] = await self.client.app_client.get_robot_parts(
+            robot_id=self.data["robot_id"]
+        )
+        filepath = call.data.get("filepath")
+        camera = call.data.get("camera")
+        component_name = call.data.get("component_name")
+        file_name = call.data.get("file_name", "camera")
+
+        if filepath is not None:
+            await self.client.data_client.file_upload_from_path(
+                filepath=filepath,
+                part_id=parts.pop().id,
+                component_name=component_name,
+            )
+        if camera is not None:
+            data = await self._get_image_from_camera(camera)
+            await self.client.data_client.file_upload(
+                part_id=parts.pop().id,
+                component_name=component_name,
+                file_name=file_name,
+                file_extension=".jpeg",
+                data=data,
+            )
+
+    async def get_classifications(self, call: ServiceCall) -> ServiceResponse:
+        """Accept input configuration to request classifications."""
+        filepath = call.data.get("filepath")
+        camera = call.data.get("camera")
+        classifier_name = call.data.get("classifier_name")
+        count = int(call.data.get("count", 2))
+        confidence_threshold = float(call.data.get("confidence_threshold", 0.6))
+
+        async with await self._get_robot_client(
+            call.data.get("robot_secret"), call.data.get("robot_address")
+        ) as robot:
+            classifier = VisionClient.from_robot(robot, classifier_name)
+            image = Image.open(filepath) if filepath is not None else None
```

Comment:
Opening a file is IO and this should not be done in the event loop to avoid blocking it. Please refactor it, that the IO is done in a function, which is run by the executor